cmd/otel-allocator/allocation/allocator.go

```diff
@@ -0,0 +1,310 @@
+func (a *allocator) unassignTargetItem(item *target.Item) {
+	collectorName := item.CollectorName
+	if collectorName == "" {
+		return
+	}
+	c, ok := a.collectors[collectorName]
+	if !ok {
+		return
+	}
+	c.NumTargets--
+	TargetsPerCollector.WithLabelValues(item.CollectorName, a.strategy.GetName()).Set(float64(c.NumTargets))
+	delete(a.targetItemsPerJobPerCollector[item.CollectorName][item.JobName], item.Hash())
+	if len(a.targetItemsPerJobPerCollector[item.CollectorName][item.JobName]) == 0 {
+		delete(a.targetItemsPerJobPerCollector[item.CollectorName], item.JobName)
+	}
+	item.CollectorName = ""
+}
+
+// removeTargetItem removes the target item from its Collector.
+func (a *allocator) removeTargetItem(item *target.Item) {
+	a.unassignTargetItem(item)
+	delete(a.targetItems, item.Hash())
+}
+
+// removeCollector removes a Collector from the allocator.
+func (a *allocator) removeCollector(collector *Collector) {
+	delete(a.collectors, collector.Name)
+	// Remove the collector from any target item records
+	for _, targetItems := range a.targetItemsPerJobPerCollector[collector.Name] {
+		for targetHash := range targetItems {
+			a.targetItems[targetHash].CollectorName = ""
+		}
+	}
+	delete(a.targetItemsPerJobPerCollector, collector.Name)
+	TargetsPerCollector.WithLabelValues(collector.Name, a.strategy.GetName()).Set(0)
+}
+
+// addCollectorTargetItemMapping keeps track of which collector has which jobs and targets
+// this allows the allocator to respond without any extra allocations to http calls. The caller of this method
+// has to acquire a lock.
+func (a *allocator) addCollectorTargetItemMapping(tg *target.Item) {
+	if a.targetItemsPerJobPerCollector[tg.CollectorName] == nil {
+		a.targetItemsPerJobPerCollector[tg.CollectorName] = make(map[string]map[string]bool)
+	}
+	if a.targetItemsPerJobPerCollector[tg.CollectorName][tg.JobName] == nil {
+		a.targetItemsPerJobPerCollector[tg.CollectorName][tg.JobName] = make(map[string]bool)
+	}
+	a.targetItemsPerJobPerCollector[tg.CollectorName][tg.JobName][tg.Hash()] = true
+}
+
+// handleCollectors receives the new and removed collectors and reconciles the current state.
+// Any removals are removed from the allocator's collectors. New collectors are added to the allocator's collector map.
+// Finally, update all targets' collectors to match the consistent hashing.
+func (a *allocator) handleCollectors(diff diff.Changes[*Collector]) {
+	// Clear removed collectors
+	for _, k := range diff.Removals() {
+		a.removeCollector(k)
+	}
+	// Insert the new collectors
+	for _, i := range diff.Additions() {
+		a.collectors[i.Name] = NewCollector(i.Name, i.NodeName)
+	}
+
+	// Set collectors on the strategy
+	a.strategy.SetCollectors(a.collectors)
+
+	// Re-Allocate all targets
+	assignmentErrors := []error{}
+	for _, item := range a.targetItems {
+		err := a.addTargetToTargetItems(item)
+		if err != nil {
+			assignmentErrors = append(assignmentErrors, err)
+			item.CollectorName = ""
```

Comment:
Why do we do this here?

Reply:
Getting rid of dubious test failures, mostly. One of the problems with having `CollectorName` be an attribute of a target is that it's not clear if targets supplied in `SetTargets` can have that attribute set, and what that means. The real target pipeline never sets this, but some tests sneakily do. I set it here instead of fixing those tests, because I really wanted my changes to pass existing tests unmodified.

This is one of the things I want to fix in a follow-up.